v2/yarpcrouter/router_test.go

```diff
@@ -101,6 +101,128 @@ func TestMapRouter(t *testing.T) {
 	}
 }
 
+func TestNewMapRouterWithEncodingProcedures_HappyCase(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	fooHandler := yarpctest.NewMockUnaryEncodingHandler(mockCtrl)
+	fooHandler.EXPECT().Handle(gomock.Any(), gomock.Any()).Return(nil, nil)
+	foo := yarpc.NewUnaryEncodingHandlerSpec(fooHandler)
+	fooCodec := yarpctest.NewMockInboundCodec(mockCtrl)
+	fooCodec.EXPECT().Decode(gomock.Any()).Return(nil, nil)
+	fooCodec.EXPECT().Encode(gomock.Any()).Return(nil, nil)
+
+	encodingProcedures := []yarpc.EncodingProcedure{
+		{
+			Name:        "happy_case",
+			HandlerSpec: foo,
+			Codec:       fooCodec,
+		},
+	}
+
+	procedures, _ := EncodingToTransportProcedures(encodingProcedures)
+	m := NewMapRouter("myservice", procedures)
+	transportProcedures := m.Procedures()
+
+	_, _, err := transportProcedures[0].HandlerSpec.Unary().Handle(context.TODO(), nil, nil)
+	assert.NoError(t, err)
+}
+
+func TestNewMapRouterWithEncodingProcedures_DecodeError(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	barHandler := yarpctest.NewMockUnaryEncodingHandler(mockCtrl)
+	bar := yarpc.NewUnaryEncodingHandlerSpec(barHandler)
+	barCodec := yarpctest.NewMockInboundCodec(mockCtrl)
+	barCodec.EXPECT().Decode(gomock.Any()).Return(nil, fmt.Errorf("some_error"))
+
+	encodingProcedures := []yarpc.EncodingProcedure{
+		{
+			Name:        "happy_case",
+			HandlerSpec: bar,
+			Codec:       barCodec,
+		},
+	}
+	procedures, err := EncodingToTransportProcedures(encodingProcedures)
+	assert.NoError(t, err)
+	m := NewMapRouter("myservice", procedures)
+	transportProcedures := m.Procedures()
+
+	_, _, err = transportProcedures[0].HandlerSpec.Unary().Handle(context.TODO(), nil, nil)
+	assert.Error(t, err)
+}
+
+func TestNewMapRouterWithEncodingProcedures_HandlerError(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	bazHandler := yarpctest.NewMockUnaryEncodingHandler(mockCtrl)
+	bazHandler.EXPECT().Handle(gomock.Any(), gomock.Any()).Return(nil, fmt.Errorf("some_error"))
+	baz := yarpc.NewUnaryEncodingHandlerSpec(bazHandler)
+	bazCodec := yarpctest.NewMockInboundCodec(mockCtrl)
+	bazCodec.EXPECT().Decode(gomock.Any()).Return(nil, nil)
+	bazCodec.EXPECT().Encode(gomock.Any()).Return(nil, nil)
+
+	encodingProcedures := []yarpc.EncodingProcedure{
+		{
+			Name:        "happy_case",
+			HandlerSpec: baz,
+			Codec:       bazCodec,
+		},
+	}
+	procedures, err := EncodingToTransportProcedures(encodingProcedures)
+	assert.NoError(t, err)
+	m := NewMapRouter("myservice", procedures)
+	transportProcedures := m.Procedures()
+
+	_, _, err = transportProcedures[0].HandlerSpec.Unary().Handle(context.TODO(), nil, nil)
+	assert.Error(t, err)
+}
+
+func TestNewMapRouterWithEncodingProcedures_EncodeError(t *testing.T) {
```

Comment:
it looks like all of these tests follow a similar pattern. Is it possible to generalize this file into a table test with subtests?
https://blog.golang.org/subtests